File: go-leetcode/array/56.合并区间.go
package leetcode

import "sort"

func merge(intervals [][]int) [][]int {
	result := [][]int{}

	// 排序
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	for i := 0; i < len(intervals); i++ {
		if i == 0 {
			result = append(result, intervals[i])
		} else if result[len(result)-1][1] < intervals[i][0] {
			result = append(result, intervals[i])
		} else {
			arr := result[len(result)-1]
			result = result[:len(result)-1] // pop
			newEnd := arr[1]
			if intervals[i][1] > newEnd {
				newEnd = intervals[i][1]
			}
			result = append(result, []int{arr[0], newEnd})
		}
	}

	return result
}

File: python-leetcode/array/56.合并区间.py
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        result = []

        # 排序
        intervals.sort(key=lambda x: x[0])

        for i in range(len(intervals)):
            if i == 0:
                result.append(intervals[i])
            elif result[-1][1] < intervals[i][0]:
                result.append(intervals[i])
            else:
                arr = result.pop()
                result.append([arr[0], max(arr[1], intervals[i][1])])

        return result
